### ISSUE TITLE
NNBD migration tool does not handle hard contravariant edges.

### ISSUE THREAD

**Issue by MichaelRFairhurst** [dart-lang/sdk#38699]
Example 1:

```dart
class A {                                                                        
  f(int x) {}                                                                    
}                                                                                
                                                                                 
class B extends A {                                                              
  @override                                                                      
  f(int x) {                                                                     
    x + 1;                                                                       
  }                                                                              
}                                                                                
                                                                                 
main() {                                                                         
  A().f(null);                                                                   
}
```

Example 2 (sourced from discussion in https://dart-review.googlesource.com/c/sdk/+/115766)

```
Iterable<int> f(List<int> x) => x.map((y) => g(y));                            
int g(int x) => x + 1;

main() {                                                                         
  f([null]);                                                                     
} 
```

The first example will result in a non-nullable parameter `x` in `B.f`. However, it probably should instead infer it as nullable and put a bang on `x! + 1`.

In the second example, the usage of `x + 1` in `g` is also marked `int x` non-nullable, through a hard edge. This then causes `y` in `(y) => g(y)` to be marked non-nullable because it is also a hard edge. However, this introduces a cast failure, as `(int) => int` is not a subtype of `(int?) => int`, the type expected of `List<int?>.map`.

Both of these are cases where contravariance and hard edges do not play well. I think we could create similar issues for any contravariant type.

It's possible that what we need is something to the effect of a "hard contravariant edge," which marks nodes as "hard nullable," which would then be a node that would stop backpropagating hard non-nullability.

**Comment by MichaelRFairhurst** [comment]
Note #40320 has a better write-up of how this manifests.

Thanks @leafpetersen !

**Comment by leafpetersen** [comment]
One thing that's not clear to me still in my example is why the migration tool believes that the parameter to "contains" is used unconditionally and hence should be non-nullable.  It's only used in that method as an argument to `Iterable.contains` which has a nullable parameter type.

**Comment by stereotype441** [comment]
@leafpetersen I've posted a theory in #40320 as to why the migration tool believes that the parameter to "contains" is unconditionally used in a non-nullable context and hence should be non-nullable.  In short, yes, it's only used as an argument to `Iterable.contains`, however due to some pretty nasty user traps I suspect that the migration tool analyzed the SDK with the NNBD experiment disabled, so it ignored the `?` in the declaration of `Iterable.contains`, and believed that it had a non-nullable parameter type.

**Comment by leafpetersen** [comment]
> In short, yes, it's only used as an argument to `Iterable.contains`, however due to some pretty nasty user traps I suspect that the migration tool analyzed the SDK with the NNBD experiment disabled, so it ignored the `?` in the declaration of `Iterable.contains`, and believed that it had a non-nullable parameter type.

@MichaelRFairhurst and I discussed this in a chat on Friday, and I think reproduced what was going off the rails.  I suspect that you are correct that the analysis_options.yaml is the main issue.  I had remembered to turned on the experiment, but neglected to commit that change into a temporary commit to avoid losing it when I reset the changes from the migration tool, so subsequent runs were done with out it.

A secondary issue is that I edited package_config.json to change the language version so that I could analyze the results.  I believe the correct workflow for now then is:

1) Add the experiment flag to analysis_options.yaml, and **commit this as a temporary patch**
2) Make sure that package_config.json has a pre-NNBD language version
3) Run the migration tool and accept the changes
4) Remove the language version marker from package_config.json
5) Run the analyzer
6) If desired, reset the directory, add + commit any hints, restore the language version to package_config.json and GOTO 2.

Does that sound right?
